### PR TITLE
[#IOCIT-232] Upgrade Redis client for Redis 6 support

### PR DESCRIPTION
**Pull request by BurnedMarshal**
This PR depends on https://github.com/pagopa/io-spid-commons/pull/133

#### List of Changes

Upgrade the Redis Client library and remove `redis-clustr`
Upgrade `io-spid-commons` with the new version that use the new Redis client

#### Motivation and Context

In order to support the Redis 6 version is needed upgrade the redis client sdk. The new official Redis Client library support natively the cluster type connection and `redis-clustr` is no more needed. 

#### How Has This Been Tested?

Unit test

#### Types of changes

- [X] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [X] Unit test
- [ ] Integration test with Redis 4 version (cluster instance)
- [ ] Integration test with Redis 6 version (cluster instance)

### CODE REVIEW

File: src/services/__tests__/redisSessionStorage.test.ts [hunk 1]
```diff
@@ -86,35 +86,36 @@ jest.mock("../../services/tokenService", () => {
...
 const mockSadd = jest.fn();
 const mockSrem = jest.fn();
 const mockSmembers = jest
   .fn()
-  .mockImplementation((_, callback) => callback(null, [mockSessionToken]));
+  .mockImplementation(_ => Promise.resolve([mockSessionToken]));
 const mockExists = jest.fn();
 const mockSismember = jest.fn();
 const mockTtl = jest.fn();
-const mockRedisClient = createMockRedis().createClient() as RedisClient;
```
Comment (arcogabbo) [inline src/services/__tests__/redisSessionStorage.test.ts:107]: here i removed the dependency `mock-redis-client`. it's not maintained anymore

File: src/services/__tests__/redisSessionStorage.test.ts [hunk 2]
```diff
@@ -386,31 +371,31 @@ describe("RedisSessionStorage#set", () => {
       expect(mockSet.mock.calls[0][0]).toBe(
         `SESSION-${aValidUser.session_token}`
       );
-      expect(mockSet.mock.calls[0][1]).toEqual(JSON.stringify(aValidUser));
```
Comment (arcogabbo) [inline src/services/__tests__/redisSessionStorage.test.ts:390]: in the new redis sdk the value is passed as the second argument in a `setEx` method. same edit was applied for the few lines below

File: src/services/redisSessionStorage.ts [hunk 3]
```diff
@@ -975,18 +980,15 @@ export default class RedisSessionStorage extends RedisStorageUtils
...
-          )
-        ),
-        TE.foldW(TE.right, keys =>
-          pipe(
-            TE.tryCatch(() => this.redisClient.del(keys), E.toError),
-            this.integerReplyAsync()
-          )
+        [...oldSessionInfoKeys, ...oldSessionKeys, ...externalTokens],
+        TE.fromPredicate(keys => keys.length === 0, identity),
+        TE.fold(
```
Comment (arcogabbo) [inline src/services/redisSessionStorage.ts:985]: this fold was not correct looking at the old implementation logic. I proceeded to repair it and refactor this block of lines

**Comment by arcogabbo** [comment]
This PR depends on https://github.com/pagopa/io-spid-commons/pull/133

File: src/controllers/__tests__/sessionController.test.ts [hunk 4]
```diff
@@ -30,16 +30,18 @@ const mockSmembers = jest.fn();
...
+  get: mockGet,
+  mGet: mockMget,
+  sMembers: mockSmembers.mockImplementation(_ =>
+    Promise.resolve([`SESSIONINFO-${mockedUser.session_token}`])
+  ),
+  sIsMember: mockSismember,
+  ttl: mockTtl,
+  set: mockSet,
+  sRem: mockSrem,
+  setEx: mockSet
```
Comment (BurnedMarshal) [inline src/controllers/__tests__/sessionController.test.ts]: Use a specific mock function for `setEx` method.

File: src/services/__tests__/redisSessionStorage.test.ts [hunk 5]
```diff
@@ -956,36 +947,31 @@ describe("RedisSessionStorage#del", () => {
...
-      expect(mockSrem).toBeCalledWith(
+      expect(mockSrem).toHaveBeenCalledTimes(1);
+      expect(mockSrem).toHaveBeenCalledWith(
         `USERSESSIONS-${aValidUserWithExternalTokens.fiscal_code}`,
-        `SESSIONINFO-${aValidUserWithExternalTokens.session_token}`,
-        expect.any(Function)
+        `SESSIONINFO-${aValidUserWithExternalTokens.session_token}`
       );
     } else {
-      expect(mockSrem).not.toBeCalled();
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/services/__tests__/redisSessionStorage.test.ts:1075]: part of the code was using `toBe...` while the other part was using `toHaveBeen...`. While refactoring i took the opportunity to make the tests aligned. Tecnically speaking, there is no difference between `toBe...` methods and `toHaveBeen...` ones.

File: src/controllers/__tests__/sessionController.test.ts (hunk 4 above)
Comment (arcogabbo) [reply src/controllers/__tests__/sessionController.test.ts]: solved in https://github.com/pagopa/io-backend/pull/962/commits/5b8820ceb39b4348a366669e07beb1dda8c2db26

File: src/services/__tests__/redisSessionStorage.test.ts [hunk 6]
```diff
@@ -979,138 +979,131 @@ describe("RedisSessionStorage#del", () => {
...
+    mockMget.mockImplementation(_ =>
+      Promise.resolve([JSON.stringify(expectedSessionInfo)])
+    );
     const response = await sessionStorage.listUserSessions(aValidUser);
     const expectedSessionInfoKey = `SESSIONINFO-${aValidUser.session_token}`;
-    expect(mockSet.mock.calls[0][0]).toBe(expectedSessionInfoKey);
-    expect(mockSet.mock.calls[0][1]).toBe(JSON.stringify(expectedSessionInfo));
-    expect(mockSadd.mock.calls[0][0]).toBe(
-      `USERSESSIONS-${aValidUser.fiscal_code}`
+    // setEx here
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/services/__tests__/redisSessionStorage.test.ts:1002]: thank you, solved in https://github.com/pagopa/io-backend/pull/962/commits/5b8820ceb39b4348a366669e07beb1dda8c2db26

File: src/services/redisSessionStorage.ts [hunk 7]
```diff
@@ -368,7 +366,7 @@ export default class RedisSessionStorage extends RedisStorageUtils
       TE.mapLeft(_ => {
         log.warn(`Error updating USERSESSIONS Set for ${user.fiscal_code}`);
       })
-    );
+    )();
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/services/redisSessionStorage.ts]: nice catch, solved in https://github.com/pagopa/io-backend/pull/962/commits/5b8820ceb39b4348a366669e07beb1dda8c2db26

File: src/services/redisSessionStorage.ts [hunk 8]
```diff
@@ -331,14 +331,12 @@ export default class RedisSessionStorage extends RedisStorageUtils
...
-    const tokens: ReadonlyArray<string> = R.collect(
+    const tokens: Array<string> = R.collect(
       (_, { prefix, value }) => `${prefix}${value}`
     )(this.getUserTokens(user));
 
     const deleteTokensPromiseV2 = await pipe(
-      // TODO: Check types for tokens array on del command
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      TE.tryCatch(() => this.redisClient.del(...(tokens as any)), E.toError),
+      TE.tryCatch(() => this.redisClient.del(tokens), E.toError),
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/services/redisSessionStorage.ts]: solved in https://github.com/pagopa/io-backend/pull/962/commits/5b8820ceb39b4348a366669e07beb1dda8c2db26

File: src/utils/redis.ts [hunk 9]
```diff
@@ -1,16 +1,25 @@
...
+): Promise<redis.RedisClientType> {
   const redisUrlOrDefault = redisUrl || "redis://redis";
   log.info("Creating SIMPLE redis client", { url: redisUrlOrDefault });
-  return redis.createClient(redisUrlOrDefault);
+  const redisClient = redis.createClient<
+    redis.RedisDefaultModules,
+    Record<string, never>,
+    Record<string, never>
+  >({
+    url: redisUrlOrDefault
```
Comment (BurnedMarshal) [inline src/utils/redis.ts:19]: The variable is now an host. Must be handled as url instead to avoid an exception

**Pull request merged**